### ISSUE TITLE
`-Werror` considered harmful

### ISSUE THREAD

**Issue by jktjkt** [onqtam/doctest#774]
**TL;DR**: please consider just enabling `-Werror` in your CI, but do *not* do that in CMake so that people who are building your stuff *see* the warnings and can decide whether to *act* on them. Chances are that they're using a different toolchain from you.

Every now and then, a new compiler adds a new type of warning. The most recent example which hit us was clang's very pedantic idea about *any* C-style buffers being considered unsafe (#766). These warnings are of course *not* enabled by default, not even by `-Wall` or `-Wextra` because they flag some perfectly valid code. Indeed, upstream itself [considers these warnings to be useful for projects which are willing to switch to an improved, hardened memory model](https://discourse.llvm.org/t/rfc-c-buffer-hardening/65734#programming-model-4). As such, these warnings are useful tools, but they are *not* recommended for every C++ project.

Unfortunately, doctest currently hard-enables *all* possible compiler warnings, *and* it activates them for anyone building the project by default. I think that this is a very bad idea, and I'm not alone; there's, for example, an [explicit policy in Gentoo](https://devmanual.gentoo.org/ebuild-writing/common-mistakes/index.html#-werror-compiler-flag-not-removed), and [this blog post makes an excellent point about its harmfulness](https://embeddedartistry.com/blog/2017/05/22/werror-is-not-your-friend/), while also pointing out how one can keep all the benefits.
